### ISSUE TITLE
Gradients are wrong for is_train=False

### ISSUE THREAD

**Issue by wielandbrendel** [apache/incubator-mxnet#7189]
I am trying to synthesise adversarial examples for which I need to compute gradients during test time. However, for the testing phase the gradients seem to be wrong. Here is a short example:

```python
import mxnet as mx
import numpy as np

# Load the pre-trained model
prefix = "Resnet/resnet-50"
model = mx.model.FeedForward.load(prefix, 0, ctx=mx.gpu(), numpy_batch_size=1)

# mean image
mean_rgb = np.array([123.68, 116.779, 103.939])[:, None, None]

np.random.seed(22)
image = np.random.uniform(size=(224, 224, 3)) * 256
label = np.array(0)

def PreprocessImage(img, show_img=False):
    sample = np.asarray(img)
    # swap axes to make image from (224, 224, 3) to (3, 224, 224)
    sample = np.swapaxes(sample, 0, 2)
    sample = np.swapaxes(sample, 1, 2)
    # sub mean 
    normed_img = sample - mean_rgb
    normed_img = normed_img.reshape((1, 3, 224, 224))
    return normed_img

# Get preprocessed batch (single image batch)
image = PreprocessImage(image)

data_sym = model.symbol.get_internals()['data']
logits_sym = model.symbol.get_children()[0]
label_sym = mx.symbol.Variable('label')

data_array = mx.nd.array(image, ctx=mx.cpu())
label_array = mx.nd.array(label[np.newaxis], ctx=mx.cpu())

arg_params = model.arg_params
aux_params = model.aux_params

arg_params[data_sym.name] = data_array
arg_params[label_sym.name] = label_array

grad_array = mx.nd.zeros(image.shape, ctx=mx.cpu())
grad_map = {data_sym.name: grad_array}

loss_sym = mx.symbol.softmax_cross_entropy(logits_sym, label_sym)

gmodel = loss_sym.bind(
    ctx=mx.cpu(),
    args=arg_params,
    args_grad=grad_map,
    grad_req='write',
    aux_states=aux_params)
gmodel.forward(is_train=False)
gmodel.backward([
    mx.nd.array(np.array([1]))
])

print('Average gradient: ', np.mean(grad_array.asnumpy()))
```
Setting `is_train=False` yields gradients that are either NaN or very large (with entries on the order of 1e18). Setting `is_train=True` yields the correct (but stochastic) gradients (with entries in the order of 1e-13).

I am not yet very experienced with MXNet. Am I making a mistake?

**Comment by wielandbrendel** [comment]
Is there any more information I can provide to help solve the problem?

**Comment by edmBernard** [comment]
can this be related with the new autograd function ?

**Comment by edmBernard** [comment]
I don't think you can compute gradient in non train mod.
The back propagation is only compute for train purpose

**Comment by wielandbrendel** [comment]
@edmBernard Thanks for your response! So there is no way to compute gradients while keeping dropout and batchnorm in testing mode?

**Comment by edmBernard** [comment]
I don't know for batch norm.
According to this issue https://github.com/dmlc/mxnet/issues/7200 it's seem it's not possible for dropout
(I haven't test myself)

**Comment by edmBernard** [comment]
I find this issue that explain how to enable dropout in prediction https://github.com/dmlc/mxnet/issues/3821

**Comment by wielandbrendel** [comment]
Enabling dropout in prediction is easy but disabling dropout while computing gradients seems not to be working.

**Comment by piiswrong** [comment]
https://github.com/apache/incubator-mxnet/pull/7303

you can use backward(is_train=False) after this PR.
batchnorm only works for is_train=False if you are not using cudnn

**Comment by wielandbrendel** [comment]
Thanks @piiswrong ! Looking forward to try the PR.

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!
Also, do please check out our [forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) for general "how-to" questions.
